src/stan/lang/grammars/semantic_actions_def.cpp

```diff
@@ -805,31 +795,26 @@ namespace stan {
     validate_void_return_allowed_f;
 
     void identifier_to_var::operator()(const std::string& name,
-                                       const var_origin& origin,
+                                       const scope& var_scope,
                                        variable& v,  bool& pass,
                                        const variable_map& vm,
                                        std::ostream& error_msgs) const {
       // validate existence
       if (!vm.exists(name)) {
-        pass = false;
+        pass = false;   // MM: no error message?  see validate_assignment below
         return;
       }
-      // validate origin
-      var_origin lhs_origin = vm.get_origin(name);
-      origin_block lhs_block = lhs_origin.program_block_;
-      if (lhs_block != local_origin
-          && lhs_block != origin.program_block_) {
+
+      // validate scope matches declaration scope or is nested in
+      scope lhs_origin = vm.get_scope(name);
+      if (lhs_origin.program_block() != var_scope.program_block()
+          && lhs_origin.program_block() != local_origin) {  // MM is_local()
         pass = false;
         return;
       }
       // enforce constancy of function args
-      if (lhs_block == function_argument_origin
-          || lhs_block == function_argument_origin_lp
-          || lhs_block == function_argument_origin_rng
-          || lhs_block == void_function_argument_origin
-          || lhs_block == void_function_argument_origin_lp
-          || lhs_block == void_function_argument_origin_rng) {
-        pass = false;
+      if (lhs_origin.fun()) {
+        pass = false;   // MM: no error message?  see validate_assignment below
```

Comment:
Same comment.  Often they can't print messages because the failure will simply cause backtracking.  So be careful.  And feel free to doc this kind of behavior (just without the `MM` or questions)